File: Algorithms/Spiral Matrix.java
//your code 
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        ArrayList<Integer> list = new ArrayList<>();
        int dir = 0, top = 0, left = 0, bottom = matrix.length - 1, right = matrix[0].length - 1;

        while (top <= bottom && left <= right) {
            // Moving left to right
            if (dir == 0) {
                for (int i = left; i <= right; i++) {
                    list.add(matrix[top][i]);
                }
                top++; // move the top boundary down
                dir = 1;
            }
            // Moving top to bottom
            else if (dir == 1) {
                for (int i = top; i <= bottom; i++) {
                    list.add(matrix[i][right]);
                }
                right--; // move the right boundary left
                dir = 2;
            }
            // Moving right to left
            else if (dir == 2) {
                for (int i = right; i >= left; i--) {
                    list.add(matrix[bottom][i]);
                }
                bottom--; // move the bottom boundary up
                dir = 3;
            }
            // Moving bottom to top
            else if (dir == 3) {
                for (int i = bottom; i >= top; i--) {
                    list.add(matrix[i][left]);
                }
                left++; // move the left boundary right
                dir = 0; // reset the direction to start again
            }
        }

        return list;
    }
}

//driver code
import java.util.List;

public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Example matrix input
        int[][] matrix = {
            {1, 2, 3},
            {4, 5, 6},
            {7, 8, 9}
        };
        
        // Call the spiralOrder function and store the result
        List<Integer> result = solution.spiralOrder(matrix);
        
        // Print the result
        System.out.println("Spiral Order: " + result);
    }
}

File: Algorithms/Spiral_matrix.cpp
#include <iostream>
#include <vector>
using namespace std;

void printSpiralMatrix(const vector<vector<int>>& matrix) {
    if (matrix.empty()) return;

    int top = 0, bottom = matrix.size() - 1;
    int left = 0, right = matrix[0].size() - 1;

    while (top <= bottom && left <= right) {
        // Traverse from left to right
        for (int i = left; i <= right; i++) {
            cout << matrix[top][i] << " ";
        }
        top++;

        // Traverse from top to bottom
        for (int i = top; i <= bottom; i++) {
            cout << matrix[i][right] << " ";
        }
        right--;

        // Traverse from right to left
        if (top <= bottom) {
            for (int i = right; i >= left; i--) {
                cout << matrix[bottom][i] << " ";
            }
            bottom--;
        }

        // Traverse from bottom to top
        if (left <= right) {
            for (int i = bottom; i >= top; i--) {
                cout << matrix[i][left] << " ";
            }
            left++;
        }
    }
}

int main() {
    int rows, cols;
    cout << "Enter the number of rows and columns: ";
    cin >> rows >> cols;

    vector<vector<int>> matrix(rows, vector<int>(cols));
    cout << "Enter the elements of the matrix:\n";
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            cin >> matrix[i][j];
        }
    }

    cout << "Spiral Order of the matrix:\n";
    printSpiralMatrix(matrix);
    cout << endl;

    return 0;
}
